crates/autopilot/src/domain/competition/participant.rs

```diff
@@ -42,51 +26,29 @@ impl<T> Participant<T> {
     }
 }
 
-impl Participant<Unscored> {
-    pub fn new(solution: Solution, driver: Arc<infra::Driver>) -> Self {
-        Self {
-            solution,
-            driver,
-            state: Unscored,
-        }
-    }
+impl<State> state::WithState for Participant<State> {
+    type State = State;
+    type WithState<NewState> = Participant<NewState>;
 
-    pub fn with_score(self, score: Score) -> Participant<Scored> {
+    fn with_state<NewState>(self, state: NewState) -> Self::WithState<NewState> {
```

Comment:
IMO having the state as a type parameter only makes sense if we keep following the type state pattern. This would mean a generic `with_state` function that allows any state transition is not permissible. If we prefer generic state transitions for whatever reason the internal state might as well become an enum and not a type parameter.